### DISCUSSION TITLE
Question about the treatment of TPROC-H test exceptions and computation of geometric mean

### DISCUSSION THREAD

**Discussion by joshuajin** [TPC-Council/HammerDB#525] [General]
During my single stream TPROC-H test (with a scale factor of 3000) in an AWS RDS cloud environment for SQL Server, it triggered SQL Server to failover. The database instance has 512GB of RAM, with a maximum IOPS of 64,000 or a maximum throughput of 1,000 MB/s. After capping the SQL Server max memory at 445GB (based on Amazon RDS BKM), the failover still occurs. Apart from the root cause of the SQL Server failover, I have two questions here and would like to get your answers or confirmations. 

1. How HammerDB TPROC-H Client deals with the exception when the client has lost its connection to the underlying database?  From the log file, I see that “The connection is broken and recovery is not possible. The connection is marked by the client driver as unrecoverable. No attempt was made to restore the connection.” Does that mean that the HammerDB TPROC-H client stopped the session? 

2. From the HammerDB GUI and the log file, I saw HammerDB TPROC-H client marked all un-executed queries as completed in less than one second. It calculated the Geometric Mean based on all successfully completed queries. In the following examples from the log file, the Query 14 and Query 2 finished successfully. Query 9 Finished in 3731.089 second but it did not get included in the calculation of Geometric Mean since it the Query 9 triggered a SQL Server Failover. Based on the successfully completed Query 14 and 2, the geometric mean calculation is 577 that matches the HammerDB TPROC-H client calculation. For my understand, I verified the Geometric Mean calculation in Excel, see screenshot. 

So, this means that when TPROC-H encounters a broken and unrecoverable connection (like a SQL Server failover event), the client will stop, the TPROC-H geometric calculation will be based on ONLY the queries that have been completed successfully in the query set (22 queries). Please confirm.

Log file content:
Hammerdb Log @ Sun Mar 12 11:15:46 MST 2023
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
Vuser 1:Executing Query 14 (1 of 22)
Vuser 1:query 14 completed in 2332.211 seconds
Vuser 1:Executing Query 2 (2 of 22)
Vuser 1:query 2 completed in 142.598 seconds
Vuser 1:Executing Query 9 (3 of 22)
Vuser 1:[Microsoft][ODBC Driver 17 for SQL Server]TCP Provider: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond. [Microsoft][ODBC Driver 17 for SQL Server]Communication link failure (executing the statement)
Vuser 1:query 9 completed in 3731.089 seconds
Vuser 1:Executing Query 20 (4 of 22)
Vuser 1:[Microsoft][ODBC Driver 17 for SQL Server]The connection is broken and recovery is not possible. The connection is marked by the client driver as unrecoverable. No attempt was made to restore the connection. (preparing statement)
Vuser 1:query 20 completed in 0.002 seconds
Vuser 1:Executing Query 6 (5 of 22)
Vuser 1:[Microsoft][ODBC Driver 17 for SQL Server]The connection is broken and recovery is not possible. The connection is marked by the client driver as unrecoverable. No attempt was made to restore the connection. (preparing statement)
Vuser 1:query 6 completed in 0.002 seconds
Vuser 1:Executing Query 17 (6 of 22)
Vuser 1:[Microsoft][ODBC Driver 17 for SQL Server]The connection is broken and recovery is not possible. The connection is marked by the client driver as unrecoverable. No attempt was made to restore the connection. (preparing statement)
Vuser 1:query 17 completed in 0.001 seconds
…
…
Vuser 1:query 7 completed in 0.001 seconds
Vuser 1:Executing Query 12 (22 of 22)
Vuser 1:[Microsoft][ODBC Driver 17 for SQL Server]The connection is broken and recovery is not possible. The connection is marked by the client driver as unrecoverable. No attempt was made to restore the connection. (preparing statement)
Vuser 1:query 12 completed in 0.001 seconds
Vuser 1:Completed 1 query set(s) in 6206 seconds
Vuser 1:Geometric mean of query times returning rows (2) is 576.68763

[Geomean_sample_tproc-h.pdf](https://github.com/TPC-Council/HammerDB/files/10973263/Geomean_sample_tproc-h.pdf)

Thank you, 
Joshua Jin

**Comment by sm-shaw** [comment]
For this example 2332.211 * 142.598 = 332568.624178 and square root of 332568.624178 = 576.68 which is our geomean result, so yes to confirm the geomean is calculated only for the successful* queries. 

*As a further qualification, the geomean is only calculated for the successful queries that return rows, so we can have a successful query that returns no rows (but not for TPROC-H), however HammerDB only inputs the values for queries where the rowcount is greater than 0 - this is intentionally done to prevent inputting values such as "query 12 completed in 0.001 seconds" that would skew the overall result. 

This means we should be looking for a result of a geomean of 22 successful queries for comparative purposes. **Where queries are failing as in this case, the geomean will not provide an accurate comparison of the overall result.** 

As further background, initially we calculated something similar to the QphH value, however the TPC requested that this be removed as it did not provide an accurate comparison to a fully audited TPC-H result - and this is valid in areas such as load times and partitioning datasets etc. Therefore, we used the Geomean calculation as an alternative as it was used by Oracle for their cloud analytic queries on a similar dataset as described here https://www.hammerdb.com/docs3.3/ch10s04.html - this value can provide a single datapoint to compare the queries against different databases at the same scale factor. 

Additional points to be aware of with TPROC-H:

- Set the Scale Factor for the Driver

Make sure you set the scale factor correctly for the schema size you are testing against in the driver script (as well as the build) to make sure the query set runs correctly. We are working on setting this automatically in a future release https://github.com/TPC-Council/HammerDB/issues/23 as this is a common issue that can sometimes catch people. 

- No Monitor VU meaning some manual calculation for Throughput tests (potential enhancement)

We also don't have a monitor virtual user https://github.com/TPC-Council/HammerDB/issues/24 for TPROC-H so if doing multiple query streams you need to manually take the slowest stream for comparison. A monitor VU would be able to do this automatically. 

- No automatic partitioning for parallel query (potential enhancement)

Similarly, at the moment we don't automatically partition the datasets, initially, as this was a paid for option in some databases and not supported in others.  However, this would also now be a valid addition as discussed here https://github.com/TPC-Council/HammerDB/discussions/369. Partitioning will enhance the performance of TPROC-H workloads done with a traditional parallel query row-oriented based approach, it is not required when running against a column store which is supported in most of the databases HammerDB supports now.  

- Insert based Columnstore loads are slow

A common question is when loading some columnstores, the load can be much slower on the database side. For this you need to create the dataset first as flat files and then do a bulk load to load the data https://www.hammerdb.com/docs/ch13.html

**Comment by joshuajin** [comment]
Steve, thanks for the answer, confirmation and additional good information. (Joshua)
